### ISSUE TITLE
Iterator.sliding(int, int) needs to be optimized

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#524]
@ruslansennov you already looked at sliding. Could you take a look if it is possible to create a _real_ Iterator for sliding? There is no hurry... Currently the code looks like this:

```java
    @Override
    default Iterator<IndexedSeq<T>> sliding(int size, int step) {
        if (size <= 0 || step <= 0) {
            throw new IllegalArgumentException(String.format("size: %s or step: %s not positive", size, step));
        }
        if (!hasNext()) {
            return empty();
        } else {
            // TODO (quick and dirty, need a real Iterator here) -->
            List<T> list = List.ofAll(() -> this);
            List<IndexedSeq<T>> result = List.empty();
            while (!list.isEmpty()) {
                final Tuple2<List<T>, List<T>> split = list.splitAt(size);
                result = result.prepend(split._1.toVector());
                list = split._2.isEmpty() ? List.Nil.instance() : list.drop(step);
            }
            return result.reverse().iterator();
            // <-- TODO
        }
    }
```

What we want is sth. like this

```java
    @Override
    default Iterator<IndexedSeq<T>> sliding(int size, int step) {
        if (size <= 0 || step <= 0) {
            throw new IllegalArgumentException(String.format("size: %s or step: %s not positive", size, step));
        }
        if (!hasNext()) {
            return empty();
        } else {
            return new AbstractIterator<T>() {
                // ...
            };
        }
    }
```

Scala uses a [`GroupedIterator`](https://github.com/scala/scala/blob/v2.11.7/src/library/scala/collection/Iterator.scala#L877) which seems way too complicated to me (if we use it for sliding only).

**Comment by ruslansennov** [comment]
OK, today evening

**Comment by danieldietrich** [comment]
great!

**Comment by danieldietrich** [comment]
Fixed with #525
